Add conflict detection to apply merge strategy

fanzhangio opened the pull request:
- Add IsConflict func for Element interface
- Add conflict detection check when option is enabled in merge visitor

**What this PR does / why we need it**:
Apply is being rewritten under pkg/kubectl/apply/strategy. The new merge and replace code should check for conflicts between the last seen value and the current value, and optionally return an error if they do not match with the field and details.

**Which issue(s) this PR fixes** 
https://github.com/kubernetes/kubectl/issues/97

**Special notes for your reviewer**:
Split the issue tasks in three subtasks. This PR focus on implement a conflict detection function for Element interface, and apply to merge visitor when FailOnConflicts option enabled.

Commit:
Add conflict detection to apply merge strategy

- Add Compare interface based on visitor pattern
- Add Compare Strategy visitor implementing Compare interfaces
- Add conflict detection check when option enabled in merge visitor
- Implement recursively conflict detection by value type
- Add tests for conflicts detection in map, list and nested combined fields of remote and recorded

fanzhangio commented:
@droot

pwittrock commented on pkg/kubectl/apply/element.go:
```diff
@@ -325,6 +326,40 @@ func IsAdd(e Element) bool {
+		var remote map[string]interface{}
+		switch e.(type) {
+		case ListElement:
+			return !reflect.DeepEqual(e.(ListElement).ListElementData.GetRecordedList(), e.(ListElement).ListElementData.GetLocalList())
```
Will this work?  What if the lists have elements with lists that need to be compared?

fanzhangio replied:
Actually I wanted to have a comment here to discuss my concerns about conflict definition in ListElement. It is very different to maps, in which conflict is definite as long as same keys have different values. ListElement objects holds remote, recorded fields in form of []interface. The problem can be converted to "compare two []interface "
Also I noticed your comments in https://github.com/kubernetes/kubernetes/blob/cdecea545553eff09e280d389a3aef69e2f32bf1/pkg/kubectl/apply/strategy/merge.go#L23-L27

I want talk about some corner cases in ListElement:
1. Does it involve `literal equality` or `deep equality` ? 
case 1:
```
record:  ["message1", "message2", "message3"]
remote:  ["message1", "message2"]
Because different number of elements, so record and remote are "not equal" . Is it a "conflict" ?
```
2. Does it require`nested validation` ? 
case 2
```
record: [{"Ports": [8080, 443]} ]
remote: [{"Ports": [8080]}]
They both hold struct as element, in which values are different (remote is missing 443), is it a conflict ?
```

My PR 01fb875 is only validating if the two ListElement object are "deep equality", for example, case 1 and case 2 will be catch as 'conflict'

I think the `conflict` may be not only about objects are equality, 'conflicts' also considers if the elements of the two list (remote and record) can be merged. In this case, case1 and case2 could be `not conflict`, is it correct ?

pwittrock commented on pkg/kubectl/apply/element.go:
```diff
@@ -325,6 +326,40 @@ func IsAdd(e Element) bool {
+
+		for k, v := range record {
+			if value, ok := remote[k]; ok {
+				if !reflect.DeepEqual(v, value) {
```
I don't think this will work, the maps may have different keys.  We only want to check if they have the same keys with different values (recursively applied)

fanzhangio replied:
The algorithm is native traversing the two maps. As we only look for the same key, it does not matter which map has more keys nor which map is outside loop. Starting traversing from the record as example,  if the key is exists in remote, which means the "two maps have same keys". Then  go DeepEqual for the values (DeepEqual works for this valuation: For same type, then valuate the value, otherwise return false)

Test example: 
```
	record := make(map[string] interface{})
	remote := make(map[string] interface{})
	record["A"] = "TEXT_A"
	record["B"] = "TEXT_B"
	record["C"] = "TEXT_C"
	record["M"] = "TestMess"
	record["X"] = 64

	remote["A"] = "REMO_A"
	remote["M"] = "TestMess"
	remote["X"] = struct {
			name  string 
			key string
		      }{name: "Name", key: "Key"}

	for k, v := range record {
 		if value, ok := remote[k]; ok {
			if !reflect.DeepEqual(v, value) {
				fmt.Printf("\nFound: Key: %v, Record: %v, Remote: %v", k, v, value)
			}
		}
	}
```
Output:
```
Found: Key: A, Record: TEXT_A, Remote: REMO_A
Found: Key: X, Record: 64, Remote: {Name Key}
```

droot commented:
had a discussion with @fanzhangio about how to address the issues pointed out by @pwittrock . @fanzhangio is going to update the PR with new approach where doing the comparison on the basis of visitor pattern inspired from merge code.

fanzhangio commented:
I figured out a new approach which is based on visitor pattern and updated PR, in which 
1) Add `Compare(Strategy)` interface for Element, and updated all implements correspondingly. 
2) Add `compareStrategy` visitor and impl for CompareMap, CompareList recursively check
3) In merge visitor (MergeMap, MergeList, etc. ) added conflict detection block with compare visitor
4) (WIP) still needs to figure out the best way of recursive conflicts check logic in CompareMap
and CompareList.
@droot  @pwittrock  @shiywang @ericchiang

anfernee commented on pkg/kubectl/apply/element.go:
```diff
@@ -36,6 +36,9 @@ type Element interface {
 	Merge(Strategy) (Result, error)
 
+	// Compare detects conflicts between recorded and remote values.
+	Compare(Strategy) bool
```
i am not familiar with the code, but this doesn't seem good. Compare should have it's own strategy right?

anfernee commented on pkg/kubectl/apply/visitor.go:
```diff
@@ -34,6 +34,18 @@ type Strategy interface {
 	MergeEmpty(EmptyElement) (Result, error)
+
+	// Compare is invoked by element when conflict detection enabled
+	Compare
```
here it is. It doesn't make sense for `Strategy` to embed `Compare`. The `Compare` method at the beginning should take this `Compare` interface, not `Strategy`

anfernee commented on pkg/kubectl/apply/strategy/merge_visitor.go:
```diff
@@ -146,3 +156,23 @@ func (v mergeStrategy) MergeEmpty(diff apply.EmptyElement) (apply.Result, error)
+
+func (v mergeStrategy) CompareType(diff apply.TypeElement) bool {
+	return v.strategic.compare.CompareType(diff)
+}
```
you don't need all this if `Strategy` doesn't include `Compare`

anfernee commented on pkg/kubectl/apply/strategy/compare_visitor.go:
```diff
@@ -0,0 +1,157 @@
+	}
+}
+
+type compareStrategy struct {
```
you don't need these too. those are for strategy.

anfernee commented on pkg/kubectl/apply/strategy/merge_visitor.go:
```diff
@@ -37,6 +37,12 @@ type mergeStrategy struct {
 func (v mergeStrategy) MergeList(e apply.ListElement) (apply.Result, error) {
+	// Conflict detection enabled for merge list
+	if v.options.FailOnConflict {
+		if !v.strategic.compare.CompareList(e) {
```
`compare` can be part of `mergeStrategy`.

dougm commented on pkg/kubectl/apply/strategy/compare_visitor.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// CompareMap recursively detects conflicts between recorded and remote in MapElement.
+// Return false if conflicts deteced.
```
typo: 'deteced' -> 'detected'

dougm commented on pkg/kubectl/apply/strategy/compare_visitor.go:
```diff
@@ -0,0 +1,157 @@
+// compareList traverses record and remote list for conflicts
+func (v compareStrategy) compareMapList(record, remote []interface{}, keys apply.MergeKeys) bool {
+	list1, list2 := sort(record, remote)
+	if len(list1) == 0 || len(list2) == 0 {
```
Could check `len()` before calling `sort`.

droot commented:
cc @mengqiy @apelisse

anfernee commented:
/ok-to-test

fanzhangio commented:
/retest

fanzhangio commented:
/retest

fanzhangio commented:
/retest

fanzhangio commented:
/retest

fanzhangio commented:
/retest

fanzhangio commented:
/retest

fanzhangio commented on pkg/kubectl/apply/compare.go:
```diff
@@ -0,0 +1,26 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018 can not pass pull-kubernetes-verify job.
